src/main/scala/is/hail/sparkextras/ContextRDD.scala

```diff
@@ -0,0 +1,159 @@
+package is.hail.sparkextras
+
+import is.hail.io._
+import is.hail.expr.types._
+import is.hail.utils._
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.storage._
+import scala.reflect.ClassTag
+
+object ContextRDD {
+  def empty[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.emptyRDD[C => Iterator[T]], mkc)
+
+  // this one weird trick permits the caller to specify T without C
+  sealed trait Empty[T] {
+    def apply[C <: AutoCloseable](
+      sc: SparkContext,
+      mkc: () => C
+    )(implicit tct: ClassTag[T]
+    ): ContextRDD[C, T] = empty[C, T](sc, mkc)
+  }
+  private[this] object emptyInstance extends Empty[Nothing]
+  def empty[T] = emptyInstance.asInstanceOf[Empty[T]]
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]]
+  ): ContextRDD[C, T] = {
+    assert(xs.length > 1)
+    val mkc = xs(0).mkc
+    assert(xs.tail.forall(_ == mkc))
+    union(sc, xs, mkc)
+  }
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]],
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.union(xs.map(_.rdd)), mkc)
+
+  def weaken[C <: AutoCloseable, T: ClassTag](
+    rdd: RDD[T],
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(rdd.mapPartitions(it => Iterator.single((ctx: C) => it)), mkc)
+
+  type ElementType[C, T] = C => Iterator[T]
+}
+
+class ContextRDD[C <: AutoCloseable, T: ClassTag](
+  val rdd: RDD[C => Iterator[T]],
+  val mkc: () => C
+) extends Serializable {
+  def run: RDD[T] =
+    rdd.mapPartitions { part => using(mkc()) { cc => part.flatMap(_(cc)) } }
+
+  private[this] def inCtx[U: ClassTag](
+    f: C => Iterator[U]
+  ): Iterator[C => Iterator[U]] = Iterator.single { (ctx: C) => f(ctx) }
+
+  private[this] def withoutContext[U: ClassTag](
+    f: Iterator[T] => Iterator[U]
+  ): ContextRDD[C, U] = new ContextRDD(rdd.map(_.andThen(f)), mkc)
```

Comment:
Isn't this literally mapPartitions?

Reply:
If multiple elements exist in a partition's iterator, this preserves it. I have avoided collapsing multiple elements unless it's necessary (as in `mapPartitions` where it seems like one should really see everything in the partition, so I `it.flatMap(_(ctx))` to get every element in the partition).

I was/am not confident in how having multiple iterators in a single partitions will work, so I took the conservative route of avoiding collapsing them unless the user really asked for it (as in `mapPartitions`).

Reply:
Ideally I wouldn't have to work with the RDD abstraction, but, 🤷‍♀️, we'll get there some day.